crates/binstalk-downloader/src/gh_api_client.rs

```diff
@@ -0,0 +1,376 @@
+    pub async fn has_release_artifact(
+        &self,
+        GhReleaseArtifact {
+            release,
+            artifact_name,
+        }: GhReleaseArtifact,
+    ) -> Result<HasReleaseArtifact, GhApiError> {
+        enum Failure {
+            Error(GhApiError),
+            RateLimit { retry_after: Instant },
+        }
+
+        let once_cell = self.0.release_artifacts.get(release.clone());
+        let res = once_cell
+            .get_or_try_init(|| {
+                Box::pin(async {
+                    use request::FetchReleaseRet::*;
+
+                    {
+                        let mut guard = self.0.retry_after.lock().unwrap();
+
+                        if let Some(retry_after) = *guard {
+                            if retry_after.elapsed().is_zero() {
+                                return Err(Failure::RateLimit { retry_after });
+                            } else {
+                                // Instant retry_after is already reached.
+                                *guard = None;
+                            }
+                        };
+                    }
+
+                    match request::fetch_release_artifacts(
+                        &self.0.client,
+                        release,
+                        self.0.auth_token.as_deref(),
+                    )
+                    .await
+                    {
+                        Ok(ReleaseNotFound) => Ok::<_, Failure>(None),
+                        Ok(Artifacts(artifacts)) => Ok(Some(artifacts)),
+                        Ok(ReachedRateLimit { retry_after }) => {
+                            let retry_after = retry_after.unwrap_or(DEFAULT_RETRY_DURATION);
+
+                            let now = Instant::now();
+                            let retry_after = now
+                                .checked_add(retry_after)
+                                .unwrap_or_else(|| now + DEFAULT_RETRY_DURATION);
+
+                            Err(Failure::RateLimit { retry_after })
+                        }
+                        Err(err) => Err(Failure::Error(err)),
+                    }
+                })
+            })
+            .await;
+
+        match res {
+            Ok(Some(artifacts)) => {
+                let has_artifact = artifacts.contains(&artifact_name);
+                Ok(if has_artifact {
+                    HasReleaseArtifact::Yes
+                } else {
+                    HasReleaseArtifact::No
+                })
+            }
+            Ok(None) => Ok(HasReleaseArtifact::NoSuchRelease),
+            Err(Failure::RateLimit { retry_after }) => {
+                *self.0.retry_after.lock().unwrap() = Some(retry_after);
+
+                Ok(HasReleaseArtifact::RateLimit { retry_after })
+            }
+            Err(Failure::Error(err)) => Err(err),
+        }
+    }
+}
+
+#[derive(Eq, PartialEq, Copy, Clone, Debug)]
+pub enum HasReleaseArtifact {
+    Yes,
+    No,
+    NoSuchRelease,
```

Comment:
I'm not sure whether `NoSuchRelease` should be in `HasReleaseArtifact` or be an variant in `GhApiError`.

From `binstalk::BinstallError` stand point, it is easier if it is part of the error, but I'm not sure this should be counted as an error.

@passcod What do you think about this?

Reply:
> From `binstalk::BinstallError` stand point, it is easier if it is part of the error

I would take this back since treating it as an error is literally a bug that caused the auto discovery system to fail.